### PR TITLE
efficient-eagles

### PR DESCRIPTION
**Pull request by MW982**

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
No merge issues

No flake8 linting errors

File: efficient-eagles/early_internet/main/ajax_views.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from django.http import JsonResponse
+
+from main.models import Post, Comment
+
+
+def helper_vote(obj, user, vote):
+    # Adds/Removes user from field upvoted_by/downvoted_by from models Post/Comment
```
Comment (eivl) [inline efficient-eagles/early_internet/main/ajax_views.py:7]: nice to see comments, but writing a docstring is normally the best way.

File: efficient-eagles/early_internet/main/models.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        abstract = True
+
+
+class CustomUser(AbstractUser):
+    profile_img = models.ImageField(upload_to="profile_imgs/", null=True, blank=True)
+    bio = models.TextField(
+        max_length=100, default="Hi, I am a HoneyFeed User", null=True, blank=True
+    )
+
+    def __str__(self):
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:21]: nice use of `__str__`

File: efficient-eagles/early_internet/main/models.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+
+class Topic(TimeStampedModel):
+    name = models.CharField(max_length=20, default="", blank=False, unique=True)
+
+    def __str__(self):
+        return self.name
+
+
+class Post(TimeStampedModel):
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:32]: nice use of delegation. Inheritance and Django work well together. 
good use of magic methods and property

File: efficient-eagles/early_internet/main/models.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        return f"author={self.author} body={self.body} thread_level={self.thread_level}"
+
+    def save(self, *args, **kwargs):
+        if self.pk:
+            self.upvotes = len(self.upvoted_by.all())
+            self.downvotes = len(self.downvoted_by.all())
+        super(Comment, self).save(*args, **kwargs)
+
+    @classmethod
+    def preorder(cls, comment, ordered_comments):
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:105]: Very nice docstring. Makes it so much easier to understand whats going on.

File: efficient-eagles/early_internet/main/models.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            Returns:
+                None
+        """
+        comments = comment.comment_set.order_by("created")
+        if not comments:
+            return None
+
+        for thread_comment in comments:
+            ordered_comments.append(thread_comment)
+            cls.preorder(thread_comment, ordered_comments)
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:122]: clever us of recursive threading.

File: efficient-eagles/early_internet/main/models.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    @classmethod
+    def sort_comment_section(cls, post):
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:125]: Same here. nice docstring.

File: efficient-eagles/early_internet/main/models.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            Returns:
+                ordered_comments (list): List of Comment objects
+        """
+        ordered_comments = []
+        comments = Comment.objects.filter(post=post, thread_level__in=[0, 1, 2, 3, 4])
+        ordered_root_comments = comments.filter(thread_level=0).order_by("-created")
+
+        for comment in ordered_root_comments:
+            ordered_comments.append(comment)
+            cls.preorder(comment, ordered_comments)
```
Comment (eivl) [inline efficient-eagles/early_internet/main/models.py:141]: It's to me a bit unclear what the purpose of cls.preorder is here. 
A function should change a value or return a value, not both.

File: efficient-eagles/early_internet/main/views.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        return redirect('login')
+
+    def post(self, request):
+        form = PasswordChangeForm(request.user, request.POST)
+        if form.is_valid():
+            user = form.save()
+            update_session_auth_hash(request, user)  # Important!
+            return redirect('account')
+
+        return redirect('change_password')
```
Comment (eivl) [inline efficient-eagles/early_internet/main/views.py:262]: overall nice short methods and clearly created view classes.

**Pull request merged**